### PR TITLE
Pointer Arithmetic

### PR DESCRIPTION
**Pull request by RalfJung**
Various mostly pointer-related things:
* After casting a pointer to usize, permit addition and subtraction
* Reject calling a function pointer with a non-0 offset

Also, while I was at it, some more tests for thread-local storage.

### CODE REVIEW

**Comment by RalfJung** [comment]
@oli-obk I hope the way I refactored the binop code is okay; I found no nice way to expand the existing code to handle pointer-integer addition.

File: src/memory.rs [hunk 1]
```diff
@@ -470,11 +480,14 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
         }
     }
 
-    pub fn get_fn(&self, id: AllocId) -> EvalResult<'tcx, ty::Instance<'tcx>> {
-        debug!("reading fn ptr: {}", id);
-        match self.functions.get(&id) {
+    pub fn get_fn(&self, ptr: Pointer) -> EvalResult<'tcx, ty::Instance<'tcx>> {
+        if ptr.offset != 0 {
```
Comment (oli-obk) [inline src/memory.rs]: Is it already UB to offset function pointers? Or just using them?

Comment (RalfJung) [inline src/memory.rs]: I would say calling `offset` on them is UB -- and miri implements this, as the bounds-check will always fail.
Calling `wrapping_offset` OTOH must not be UB, as that's a safe function.

Comment (oli-obk) [inline src/memory.rs]: Can you actually turn a function pointer to a raw pointer in safe code? If not, we can also check for fn ptrs in `wrapping_offset`

Comment (eddyb) [inline src/memory.rs]: Yes, `main as u8` is the joke way to do this (w/ `playbot`) 😆. Casts from functions or function pointers to *any* pointers or integers are safe.

Comment (oli-obk) [inline src/memory.rs]: Ugh. ..

**Pull request merged**